<a href="https://colab.research.google.com/github/faaddhhllii/llyods_bank_prediction_churn/blob/main/Data_Understanding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Understanding

In [ ]:
# import library
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# read excel file
file_path = "/content/Customer_Churn_Data_Large.xlsx"
excel_file = pd.ExcelFile(file_path)

# read each sheet into the dataframe
df_demo = pd.read_excel(excel_file, sheet_name="Customer_Demographics")
df_trans = pd.read_excel(excel_file, sheet_name="Transaction_History")
df_service = pd.read_excel(excel_file, sheet_name="Customer_Service")
df_online = pd.read_excel(excel_file, sheet_name="Online_Activity")
df_churn = pd.read_excel(excel_file, sheet_name="Churn_Status")

# Merge all dataframes based on Customer ID
df_merged = df_demo \
            .merge(df_trans, on="CustomerID") \
            .merge(df_service, on="CustomerID") \
            .merge(df_online, on="CustomerID") \
            .merge(df_churn, on="CustomerID")

# view aggregated data
df_merged.head()

CustomerID  Age Gender MaritalStatus IncomeLevel  TransactionID  \
0           1   62      M        Single         Low           7194   
1           2   65      M       Married         Low           7250   
2           2   65      M       Married         Low           9660   
3           2   65      M       Married         Low           2998   
4           2   65      M       Married         Low           1228   

  TransactionDate  AmountSpent ProductCategory  InteractionID InteractionDate  \
0      2022-03-27       416.50     Electronics           6363      2022-03-31   
1      2022-08-08        54.96        Clothing           3329      2022-03-17   
2      2022-07-25       197.50     Electronics           3329      2022-03-17   
3      2022-01-25       101.31       Furniture           3329      2022-03-17   
4      2022-07-24       397.37        Clothing           3329      2022-03-17   

  InteractionType ResolutionStatus LastLoginDate  LoginFrequency ServiceUsage  \
0         Inquiry         Resolved    2023-10-21              34   Mobile App   
1         Inquiry         Resolved    2023-12-05               5      Website   
2         Inquiry         Resolved    2023-12-05               5      Website   
3         Inquiry         Resolved    2023-12-05               5      Website   
4         Inquiry         Resolved    2023-12-05               5      Website   

   ChurnStatus  
0            0  
1            1  
2            1  
3            1  
4            1

above are the initial 5 rows of the dataframe that have been merged

In [ ]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5204 entries, 0 to 5203
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   CustomerID        5204 non-null   int64         
 1   Age               5204 non-null   int64         
 2   Gender            5204 non-null   object        
 3   MaritalStatus     5204 non-null   object        
 4   IncomeLevel       5204 non-null   object        
 5   TransactionID     5204 non-null   int64         
 6   TransactionDate   5204 non-null   datetime64[ns]
 7   AmountSpent       5204 non-null   float64       
 8   ProductCategory   5204 non-null   object        
 9   InteractionID     5204 non-null   int64         
 10  InteractionDate   5204 non-null   datetime64[ns]
 11  InteractionType   5204 non-null   object        
 12  ResolutionStatus  5204 non-null   object        
 13  LastLoginDate     5204 non-null   datetime64[ns]
 14  LoginFrequency    5204 n

in total there are 16 features and 5204 rows of data


In [ ]:
# data description
df_merged.describe()

CustomerID          Age  TransactionID                TransactionDate  \
count  5204.000000  5204.000000    5204.000000                           5204   
mean    496.470407    43.140085    5495.988470  2022-06-29 08:30:15.219062272   
min       1.000000    18.000000    1000.000000            2022-01-01 00:00:00   
25%     244.000000    29.000000    3218.000000            2022-03-30 00:00:00   
50%     504.000000    43.000000    5490.000000            2022-06-27 00:00:00   
75%     743.000000    57.000000    7713.000000            2022-09-28 00:00:00   
max     995.000000    69.000000    9997.000000            2022-12-31 00:00:00   
std     286.681289    15.430343    2595.575295                            NaN   

       AmountSpent  InteractionID                InteractionDate  \
count  5204.000000    5204.000000                           5204   
mean    254.354940    5921.861261  2022-07-01 16:31:10.561106944   
min       5.180000    2015.000000            2022-01-01 00:00:00   
25%     128.940000    3873.250000            2022-04-07 00:00:00   
50%     255.175000    5903.000000            2022-07-02 00:00:00   
75%     378.605000    7851.000000            2022-09-25 00:00:00   
max     499.700000    9997.000000            2022-12-30 00:00:00   
std     143.329654    2332.331260                            NaN   

                       LastLoginDate  LoginFrequency  ChurnStatus  
count                           5204     5204.000000  5204.000000  
mean   2023-07-04 09:27:31.960030720       25.918332     0.205995  
min              2023-01-01 00:00:00        1.000000     0.000000  
25%              2023-04-04 00:00:00       14.000000     0.000000  
50%              2023-07-11 00:00:00       27.000000     0.000000  
75%              2023-10-01 00:00:00       38.000000     0.000000  
max              2023-12-31 00:00:00       49.000000     1.000000  
std                              NaN       14.110627     0.404466

This dataset contains 5204 transaction data with a total of 995 unique customers. The average customer age is 43 years, with an age range between 18 and 69 years. The average customer expenditure (AmountSpent) is around 254, with a minimum value of 5.18 and a maximum value of 499.7, showing considerable variation.

The frequency of customer logins was recorded on average 26 times, with a range between 1 to 49 times. In terms of churn, around 20.6% of customers churn while 79.4% still persist, so the data is classified as imbalance. Transaction activity was recorded throughout 2022, while the last login was until July 2023, which shows that there is activity data after the main transaction period.

In [ ]:
# delete id feature
df_merged.drop(["CustomerID", "TransactionID", "InteractionID"], axis=1, inplace=True)

removed all ID features because they are unique in value and can interfere with the machine learning model

In [ ]:
# check missing value
df_merged.isnull().sum()

Age                 0
Gender              0
MaritalStatus       0
IncomeLevel         0
TransactionDate     0
AmountSpent         0
ProductCategory     0
InteractionDate     0
InteractionType     0
ResolutionStatus    0
LastLoginDate       0
LoginFrequency      0
ServiceUsage        0
ChurnStatus         0
dtype: int64

All features have no missing values